wire-ios-data-model/Source/MLS/MLSService.swift

```diff
@@ -1681,7 +1681,65 @@ public final class MLSService: MLSServiceInterface {
     // MARK: - Proteus to MLS Migration
 
     public func startProteusToMLSMigration() async throws {
-        // Migrate proteus team group conversations to MLS
+        guard let context = context else {
+            assertionFailure("MLSService.context is nil")
+            return
+        }
+
+        let groupConversations = try await context.perform {
+            try ZMConversation.fetchAllTeamGroupConversations(messageProtocol: .proteus, in: context)
+        }
+        for conversation in groupConversations {
+
+            guard let qualifiedID = conversation.qualifiedID else {
+                logger.warn("skipping migration of conversation \(conversation), `qualifiedID` is `nil`")
+                assertionFailure("the group conversation has no `qualifiedID` set")
+                continue
+            }
+            let members = conversation.localParticipants.map { MLSUser(from: $0) }
+            let mlsGroupID: MLSGroupID
+
+            do {
+
+                // update message protocol to `mixed`
+                try await actionsProvider.updateConversationProtocol(qualifiedID: qualifiedID, messageProtocol: .mixed, context: context.notificationContext)
+
+                // sync the group conversation
+                try await actionsProvider.syncConversation(qualifiedID: qualifiedID, context: context.notificationContext)
+
+                guard conversation.mlsGroupID != nil else {
```

Comment:
shouldn't this be accessed from the context thread ?

Reply:
True, I restructured it: 99ca96b938afde2ea33d7a47d2f932ec82db2d62

However, I am starting to wonder if the performance of the code is still good, considering that we jump between queues every few statements.
Maybe it would be better to write synchronous methods and perform larger pieces on the sync queue.

What do you think? @David-Henner

Reply:
I don't know what's the real impact of jumping to the context queue so frequently. But maybe we should try to group operations on the same queue to reduce switching? Another solution would be to reduce the dependency on managed objects